### PR TITLE
add Node::describe()

### PR DESCRIPTION
**Pull request by cyrush**
requested by @mclarsen 

add a describe method, like r or pandas data frames provide. 

for each leaf, it provides summary stats, but also a truncated display of the values.

For larger datasets, using the describe output is helpful for debugging and understand vs wall of text from standard to yaml. 

TODO:

[x] settle on `count` vs `number_of_elements`
[x] add to python interface

### CODE REVIEW

**Comment by cyrush** [comment]
Added python support. 

For prior `count` vs `number_of_elements` quandry:

`count` is used much more widely in this context, so I think we should stick with it.

**Comment by cyrush** [comment]
@mclarsen @xjrc can you give this a quick review when you get a chance?

File: src/libs/conduit/conduit_data_array.cpp [hunk 1]
```diff
@@ -1436,6 +1511,135 @@ DataArray<T>::compact_elements_to(uint8 *data) const
...
+        if( (threshold % 2) > 0)
+        {
+            bottom++;
+        }
+
+        if(nele > 1)
+            os << "[";
+
+        bool first = true;
+        bool done  = false;
```
Comment (xjrc) [inline src/libs/conduit/conduit_data_array.cpp]: It seems to me like something weird will happen when the list is empty as the `while` loop will evaluate at least once; thus, I think this should be `bool done = nele == 0;`.

File: src/libs/conduit/conduit_data_array.cpp [hunk 2]
```diff
@@ -1436,6 +1511,135 @@ DataArray<T>::compact_elements_to(uint8 *data) const
...
+
+        if( (threshold % 2) > 0)
+        {
+            bottom++;
+        }
+
+        if(nele > 1)
+            os << "[";
+
+        bool first = true;
```
Comment (xjrc) [inline src/libs/conduit/conduit_data_array.cpp]: To reduce the number of variables that we need to keep track of, would it make sense to use `idx == 0` as the check for the first element instead of a separate `first` variable?

File: src/tests/conduit/t_conduit_array.cpp [hunk 3]
```diff
@@ -609,6 +609,60 @@ TEST(conduit_array, print_bells_and_whistles)
...
+    EXPECT_EQ(v,"[1, 2, 3, 4, 5]");
+
+    v = va_int64.to_summary_string(2);
+    std::cout << v << std::endl;
+    EXPECT_EQ(v,"[1, ..., 5]");
+
+    v = va_int64.to_summary_string(3);
+    std::cout << v << std::endl;
+    EXPECT_EQ(v,"[1, 2, ..., 5]");
+
```
Comment (xjrc) [inline src/tests/conduit/t_conduit_array.cpp:665]: I think it'd be good to add some more test cases here for the corner cases, e.g. when `threshold` is bigger than the array size, when the array is empty, etc.

File: src/tests/conduit/t_conduit_node.cpp [hunk 4]
```diff
@@ -1162,3 +1162,60 @@ TEST(conduit_node, add_child)
...
+    EXPECT_EQ(d["f/min"].to_float(),1.0);
+
+    EXPECT_EQ(d["a/max"].to_int(),5);
+    EXPECT_EQ(d["b/max"].to_int(),3);
+    EXPECT_EQ(d["c/max"].to_int(),6);
+    EXPECT_EQ(d["d/max"].to_int(),7);
+    EXPECT_EQ(d["e/max"].to_int(),12);
+    EXPECT_EQ(d["f/max"].to_float(),7.0);
+
+    EXPECT_EQ(d["g/mean"].to_float(),3.0);
```
Comment (xjrc) [inline src/tests/conduit/t_conduit_node.cpp:1209]: Just so we have test cases for some of the `int` arrays, I think adding `EXPECT_EQ(d["a/mean"].to_float(),3.0);` and `EXPECT_EQ(d["b/mean"].to_float(),2.0);` here would be beneficial.

File: src/libs/conduit/conduit_node.hpp [hunk 5]
```diff
@@ -3708,10 +3708,18 @@ class CONDUIT_API Node
...
     /// TODO: this is inefficient w/o move semantics, but is very
     /// convenient for testing and example programs.
     Node             info() const;
 
+    ///
+    /// describe() creates a node that replaces each leaf with
+    ///  descriptive statistics (count, mean, min, max) and a string repd 
+    ///  values summary
+    void             describe(Node &nres) const;
+    void             describe(const Node &opts, Node &nres) const;
```
Comment (xjrc) [inline src/libs/conduit/conduit_node.hpp:3721]: Could we also add `to_summary_string` and `to_summary_string_default` to the `conduit::Node` type? This would make it much easier to print summary information from debugging tools like `gdb`.

**Review by xjrc — CHANGES_REQUESTED** [review]
These new functions look great Cyrus; they should make debugging large Conduit trees much easier! I have a few minor questions regarding tests and interface changes, but once we hash those out I think this is good to go.

File: src/libs/conduit/conduit_node.hpp (hunk 5 above)
Comment (cyrush) [reply src/libs/conduit/conduit_node.hpp:3721]: This is an awesome idea, I will address in a new PR, here is an issue associated with this 

https://github.com/LLNL/conduit/issues/674

File: src/libs/conduit/conduit_data_array.cpp (hunk 1 above)
Comment (cyrush) [reply src/libs/conduit/conduit_data_array.cpp]: good catch, i made this change.

File: src/libs/conduit/conduit_data_array.cpp (hunk 2 above)
Comment (cyrush) [reply src/libs/conduit/conduit_data_array.cpp]: agreed, done

File: src/tests/conduit/t_conduit_array.cpp (hunk 3 above)
Comment (cyrush) [reply src/tests/conduit/t_conduit_array.cpp:665]: thanks for keeping me honest -- i added several mode

File: src/tests/conduit/t_conduit_node.cpp (hunk 4 above)
Comment (cyrush) [reply src/tests/conduit/t_conduit_node.cpp:1209]: agreed, I added mean checks for all arrays

**Review by xjrc — APPROVED** [review]
Thanks for all of the fixes!
